src/common/middleware/call-api.js

```diff
@@ -17,39 +19,71 @@ export default defaults => () => next => async action => {
   validateDefaults(defaults);
   validateSettings(settings);
 
-  const [ requestType, successType, failureType ] = settings.types;
+  const requestType = settings.types ? settings.types[0] : null;
+  const successType = settings.types ? settings.types[1] : null;
+  const failureType = settings.types ? settings.types[2] : null;
   const resource = defaults.endpoint + settings.path;
   const optionsWithCsrfToken = withCsrfToken(defaults.csrfToken, settings.options);
   const createAction = createActionWith.bind({}, action);
 
   // Immediately dispatch request action
-  next(createAction({ type: requestType }));
+  if (requestType) {
+    next(createAction({ type: requestType }));
+  }
 
-  try {
-    const response = await fetch(resource, optionsWithCsrfToken);
-    await checkStatus(response);
-    const result = await response.json();
+  return fetch(resource, optionsWithCsrfToken)
+    .then((response) => {
+      return checkStatus(response)
+        .then(() => {
+          return response.json()
+            .then((result) => {
+              if (result.status !== 'success') {
+                throw getError(response, result);
+              }
 
-    if (result.status !== 'success') {
-      throw getError(response, result);
-    }
+              if (successType) {
+                next(createAction({
+                  type: successType,
+                  payload: {
+                    response: result
+                  }
+                }));
+              }
+
+              return Promise.resolve(result);
+            });
+        })
+        .catch((error) => {
+          if (failureType) {
+            next(createAction({
+              type: failureType,
+              payload: {
+                error: error
+              },
+              error: true
+            }));
+          }
 
-    return next(createAction({
-      type: successType,
-      payload: {
-        response: result
+          error.action = action;
+
+          return Promise.reject(error);
+        });
+    })
+    .catch((error) => {
+      if (failureType) {
+        next(createAction({
```

Comment:
Stopped early-return here if `failureType` is defined. This means a promise is always returns as long as `[CALL_API]` settings are set on action.

This should make it a little easier to understand.